pwiz_tools/clean-apps.bat

```diff
@@ -16,7 +16,21 @@ for /d /r Shared\MSGraph %%d in (obj, bin) do @if exist "%%d" rmdir /s/q "%%d"
 for /d /r Shared\ProteomeDb %%d in (obj, bin) do @if exist "%%d" rmdir /s/q "%%d"
 for /d /r Shared\ProteowizardWrapper %%d in (obj, bin) do @if exist "%%d" rmdir /s/q "%%d"
 for /d /r Shared\zedgraph %%d in (obj, bin) do @if exist "%%d" rmdir /s/q "%%d"
-for /d /r Skyline %%d in (obj, bin) do @if exist "%%d" rmdir /s/q "%%d"

```

Comment:
I guess I'd have to know how expensive it is to rename that - presumably it's referenced in another tool. I was thinking just meet them where they are...

Reply:
@eduardo-proteinms Is that path used in a script? If so how hard to change?

Reply:
It can just be renamed and wont have any impact

Reply:
That simplifies things, for sure.

Reply:
Or not... I'd forgotten that this is in a forked submodule. This kind of meddling doesn't sit right with me - feels like the tail wagging the dog. I don't like changing more in a fork than is absolutely necessary, and mucking with the directory structure in an otherwise untouched region of the forked code doesn't feel right. Next time we pull in a submodule it might not even be possible to do such a thing without major surgery in the fork. My solution is clean and general.

Reply:
Well. The issue with it is that while a developer might commit something inside a bin folder, it might also still be a folder we want to clean up. So, it is actually a little better to have that end up flagged and have a developer make an explicit decision what to do about it. My Utf16ToUft8 tool is a perfect example. I hadn't even thought about what a clean operation would do to it. In the end, I agreed with Nick's idea to move it to a separate folder from where my project created Debug and Release folders and built its artifacts.

The question in Eduardo's project might be whether or not the root project has these binaries committed to "bin" or whether he did that as a convenience for us, as I did for Utf16ToUtf8.

Reply:
Yes I agree with Brendan. If the submodule has the "bin" folder then we should either:
- update the clean to avoid cleaning that (as you have done)
- move the submodule outside Skyline; submodules could probably be more simply cleaned with a `git clean` command and could live in pwiz_tools/submodules (obviously a more disruptive change without much upside).

If eduardo added "bin" as a convenience we can just drop/rename it and defer the above issue until it happens. One problem with checking binaries into the repo alongside their source code is they easily get out of sync with the source. If these are easy projects to build it might be better to check in a script that builds it if necessary, then runs it. That's obviously easier to do for a .NET project than some C project though.

Reply:
The folder in question is a pre-existing part of the project that was forked, which is why I'm resisting messing with it. Sorry I didn't make that more clear.

I can look at further tweaking the script to individually address files within the folder, to handle Brendan's scenario of a folder with a mix of checked in and temporary files.

Reply:
OK. Let's move this into CleanSkyline.bat then. Can we get away with a `git clean` though instead of this unfortunately verbose `git ls-files` inside a FOR loop? Like we use this for vendor data directories:
git clean -f -d -X pwiz\data\vendor_readers\UIMF\Reader_UIMF_Test.data > nul

It looks like quite a few people added bin/obj cleaning to CleanSkyline.bat when it was already handled recursively by clean-apps.bat. If `git clean` works it'd greatly simplify that CleanSkyline.bat file. :)

Reply:
The only problem with "git clean" is that it will blow away any new files that you meant to "git add" but hadn't gotten around to yet because you're in the middle of some new feature development, along with any local files like "SkylineTester test list.txt" which never get added to git but lives in the source code area.

Reply:
`git clean -x` will do that, but not -X. -X only cleans files in the ignore list, e.g. files that you expect the build to create.

Reply:
The problem with -X is it does blow away local preference files like .csproj.user and .DotSettings.user. And unfortunately it looks like the -e exclusion option doesn't work on -X, only -x. Darnit. We could use -x with exclusions to protect code files, but I don't like it blowing away a bunch of untracked files in there that aren't code files.

https://stackoverflow.com/questions/59816592/git-exclude-some-files-when-ran-git-clean
It doesn't look like there's a way to do it with git clean. We'd have to fall back to `git ls-files --ignored` and since we don't have xargs on Windows it'd need one of those FOR loops. But if we did come up with a list of exclusions from the ignore list (as if something like `git clean -X -e "*.user" -e "*.suo"` actually worked), we could definitely make the cleanup process easier to maintain...

Reply:
So I vote for just moving your changes to CleanSkyline.bat (as well as the recursive obj FOR loop), maybe get rid of the redundant bin/obj cleanings in CleanSkyline.bat, and call it a day.

Reply:
I think applying git clean to just the obj and bin folders is reasonable. Have a look at my latest commit and see what you think.

Reply:
Yeah, that's reasonable. It's unlikely there's going to be any files we don't want blown away in a bin/obj folder. Are you sure it's working as intended? I added a dirty repo check in tcbuild.bat in another PR, but currently there's no checking in TC that the repo doesn't get dirtied by the clean operation.

Reply:
I did manually test by adding untracked files to the bin directory in question, and its parent (not bin or obj) directory. The file in the bin directory was removed, the one in the parent was not. The tracked file in the bin directory was retained.